server/service/src/sensor/update.rs

```diff
@@ -54,6 +60,98 @@ pub fn update_sensor(
     Ok(sensor)
 }
 
+pub fn update_sensor_logs_for_breach(
+    connection: &StorageConnection,
+    breach_id: &String,
+) -> Result<Vec<TemperatureLog>, RepositoryError> {
+    let mut temperature_logs: Vec<TemperatureLog> = Vec::new();
+
+    let breach_result =
+        TemperatureBreachRowRepository::new(connection).find_one_by_id(breach_id)?;
+
+    if let Some(mut breach_record) = breach_result {
+        let is_cumulative_breach = (breach_record.r#type
+            == TemperatureBreachRowType::ColdCumulative)
+            | (breach_record.r#type == TemperatureBreachRowType::HotCumulative);
+        let logs = get_sensor_logs_for_breach(connection, breach_id)?; //sorted by date/time
+        let mut log_interval = 0;
+        let zero_time = NaiveTime::parse_from_str("00:00", "%H:%M").unwrap(); // hard-coded -> should always work!
+
+        if let Some(sensor) = SensorRowRepository::new(connection).find_one_by_id(&breach_record.sensor_id)? {
+            if let Some(interval) = sensor.log_interval {
+                log_interval = interval;
+            }
+        }
+        
+        if is_cumulative_breach {
+            // Update breach start/end from first/last logs if it has changed
+            if let Some(first_log) = logs.first() {
+                // If within log_interval seconds of the start of the day,
+                // use the start of the day as first log time
+                let mut first_breach_datetime = NaiveDateTime::new(first_log.temperature_log_row.datetime.date(), zero_time);
+                let first_log_datetime = first_breach_datetime + Duration::seconds(log_interval.into());
+
+                if first_log_datetime < first_log.temperature_log_row.datetime {
+                    first_breach_datetime = first_log.temperature_log_row.datetime;
+                }
+
+                // Use the first log time if it's either earlier than the calculated breach end
+                // or if the calculated breach start is less than the sensor log interval before the first log time
+                if (breach_record.start_datetime > first_breach_datetime) | (log_interval > 0) & (breach_record.start_datetime < first_breach_datetime - Duration::seconds(log_interval.into())) {
+                    log::info!(
+                        "Updating cumulative breach start for {:?} to {:?}",
+                        breach_record,
+                        first_breach_datetime
+                    );
+                    breach_record.start_datetime = first_breach_datetime;
+                    TemperatureBreachRowRepository::new(connection).upsert_one(&breach_record)?;
+                }
+            }
+            if let Some(last_log) = logs.last() {
```

Comment:
> and start

Sorry I missed that, isn't start date always provided under `A:` ?

Reply:
The breach start and end timestamps _are_ provided for Q-tags, so those are fine (and none of these adjustments are needed as we can just use the provided breach data as is), but not for Fridge-tags - all we get is the breach activation timestamp (which is when the breach has been going long enough to trigger when it gets to the breach threshold duration) and the breach duration, so the best we can do is to try to work out the start & end from the data that we do have:

(1) for a fridge tag without logging, we have the activation timestamp and breach duration, and a few implied temperature logs for the min/max temperatures for each day:
- the crate calculates the cumulative breach start timestamp by subtracting the breach threshold duration from the activation timestamp (or midnight if the activation timestamp is within the breach threshold duration from midnight, which indicates an ongoing breach from the previous day) [as mentioned above, this only works if there are no gaps in the cumulative breach up to the activation timestamp]
- the crate calculates the cumulative breach end timestamp by adding on the breach duration to the calculated start timestamp (or midnight if that is sooner) [and this only works if there are no gaps in the cumulative breach at all, although you could also work it out by adding the (breach duration - breach threshold duration) to the activation timestamp, and that would maybe allow for a gap before the activation timestamp]
-> there's nothing more that omSupply can do, other than assign any of the min/max temperature logs to a cumulative breach if they happen to fall within its duration
=> there are minimal temperature logs available (just daily min/max); cumulative breach start/end timestamps are only valid if it's a continuous breach; no consecutive breach data unless you say that all cumulative breaches are also consecutive ones (which may be a valid approach to take??). All of this may be moot as I understand that we're not committing to support this type of fridge tag anyway @adamdewey ?

(2) for fridge tags with logging data, we have additional temperature logs, usually at 5-minute intervals. That lets us cross-check the calculated breach data for consistency - you'd expect that any hot/cold breach temperature logs should be within the corresponding breach start/end timestamp range, and that's when you discover the discrepancies, especially if there's a gap (bearing in mind that we only get _cumulative_ breaches from the Fridge-tag data)...
-> we can use the temperature logs to correct the limitations in the breach start/end calculations:
- if the first breached temperature log is within the logging interval to midnight, then use midnight as the effective log timestamp
- if the last breached temperature log is within the logging interval from midnight, then use midnight as the effective log timestamp in the adjustments below
- if the first breached temperature log is before the breach start timestamp, then extend the breach start range to include the first log timestamp in the adjustments below
- if the last breached temperature log is after the breach end timestamp, then extend the breach range to include the last log timestamp
- if the first breached temperature log is more than the logging interval after the breach start timestamp, then reduce the breach range to start with the first log timestamp
- if the last breached temperature log is more than the logging interval before the breach end timestamp, then reduce the breach range to end with the last log timestamp
=> the first two of these are to avoid the (completely naff IMHO) situation where, for example, an all-day cumulative breach is shown as having started at 00:02 and ended at 23:57, due to the 5-minute logging interval; the others correct discrepancies in the calculated breach start/end timestamps; if all temperature logs are within the given breach range, then nothing gets changed

Reply:
To put your mind at rest, the crate itself doesn't attempt to do any data patching up - the only thing it does is to try to work out the start times of cumulative breaches for fridgetags from the breach activation time minus the breach config duration, and then adds on the breach duration to get the breach end time => those won't be accurate in some cases, but the most important thing is the breach duration itself, not the exact times when it started & finished.

I don't remember off the top of my head if that info is already in the crate docs, but I'll add it if it isn't.